### ISSUE TITLE
nn.Embedding(), max_norm is not None can`t fix the random seed on cuda: #4340

### ISSUE THREAD

**Issue by bamtercelboo** [pytorch/pytorch#4514]
The problem : [#4340](https://github.com/pytorch/pytorch/issues/4340)
I have find the bug for the problem, the bug appear when use nn.Embedding() , max_norm is not None.
if you use like this, self.embed = nn.Embedding(V, D, max_norm=None, scale_grad_by_freq=True) , this is ok, no problem, gpu and cpu both ok.
but when you use it like this, self.embed = nn.Embedding(V, D, max_norm=5, scale_grad_by_freq=True) , the random seed can not fix in gpu, cpu is ok.
so, I wonder to know how to fix the random seed when use max_norm is not None, whether there is some error in nn.Embedding().

**Comment by apaszke** [comment]
Do you have any (small) snippet that would let us reproduce what you're seeing? I'm not sure if I fully understand the problem. `nn.Embedding` is fully deterministic and has no influence on the random state.

**Comment by bamtercelboo** [comment]
@apaszke all right, the test following:

- the same demo, **max_norm is not None**, ` self.embed = nn.Embedding(V, D, max_norm=5, scale_grad_by_freq=True)`run two times on `cuda`:
first result:
![default](https://user-images.githubusercontent.com/22359268/34649069-3e3d5556-f3e2-11e7-8f0c-12ac4d71a79a.png)
second result:
![default](https://user-images.githubusercontent.com/22359268/34649076-64cbf0f6-f3e2-11e7-8854-155ca218ae6e.png)

- **two times of loss and acc is not same.**

- the same demo, **max_norm is None**, ` self.embed = nn.Embedding(V, D, max_norm=None, scale_grad_by_freq=True)`run two times on `cuda`:
first result:
![default](https://user-images.githubusercontent.com/22359268/34649096-d3924404-f3e2-11e7-94a5-d543950ca532.png)
second result: 
![default](https://user-images.githubusercontent.com/22359268/34649100-ea33b666-f3e2-11e7-9e0a-c5625b07ce81.png)

- **two times of loss and acc is same.**

- above test, when run on `cpu`, max_norm is None or max_norm is not None has **no influence** on the random state.

- above test, **nothing changed** every time, except the parameter of max_norm.

- above demo, you can see [here](https://github.com/bamtercelboo/cnn-lstm-bilstm-deepcnn-clstm-in-pytorch/blob/master/models/model_CNN.py)

- Are you sure `nn.Embedding()` has no influence on the random state when run demo **on gpu(cuda)**?

**Comment by bamtercelboo** [comment]
@apaszke and,  when set the `torch.backends.cudnn.enabled = False`, the max_norm is not None else has no influence on the random state, so I think the problem is the cudnn problem, but I can not understand why the parameter of max_norm has relation with the cudnn?

**Comment by soumith** [comment]
what is the output of `print(torch.__version__)`?

**Comment by bamtercelboo** [comment]
@soumith I once doubt the problem appear the pytorch version, I use the version is 0.3, I test it on the version 0.2, but still has the problem. so, the problem not caused by pytorch version.

**Comment by soumith** [comment]
@bamtercelboo i checked the CUDA kernel for max_norm based Embedding. It has no AtomicAdd operations and should be fully deterministic. What is likely happening is that for some reason when `max_norm=False`, by chance cudnn must be picking it's deterministic algorithms. This can happen because of subtle reasons (maybe if max_norm=False the input to Conv2d backward somewhere is non-contiguous or something...).

We will likely not take any action on this issue.
If you want to retain cudnn's determinism, we introduced a flag:

```
torch.backends.cudnn.deterministic = True
```
which makes sure CuDNN always picks up deterministic algorithm (but it will be slower).

**Comment by bamtercelboo** [comment]
@soumith @apaszke all right, I have tested,  `torch.backends.cudnn.deterministic = True`, the speed is almost same with the default, `torch.backends.cudnn.enabled = False`, the speed will be slower so much, but all of them can fix the cudnn random state. thanks.

**Comment by bamtercelboo** [comment]
@apaszke @soumith `torch.backends.cudnn.deterministic = True` is only effective on CNN, not effective on LSTM, if want to fix the lstm random state in gpu, only set `max_norm = None` until now.

**Comment by apaszke** [comment]
Note that it's not really caused by the `max_norm` flag (it's some weird interaction between cuDNN and the result of computation), so it might still get triggered under different conditions unless you use `deterministic = True`
